// victory-native/dist/cartesian/utils/transformInputData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformInputData = void 0;
const tickHelpers_1 = require("../../utils/tickHelpers");
const asNumber_1 = require("../../utils/asNumber");
const makeScale_1 = require("./makeScale");
/**
 * This is a fatty. Takes raw user input data, and transforms it into a format
 *  that's easier for us to consume. End result looks something like:
 *  {
 *    ix: [1, 2, 3], // input x values
 *    ox: [10, 20, 30], // canvas x values
 *    y: {
 *      high: { i: [3, 4, 5], o: [30, 40, 50] },
 *      low: { ... }
 *    }
 *  }
 *  This form allows us to easily e.g. do a binary search to find closest output x index
 *   and then map that into each of the other value lists.
 */
const transformInputData = ({ data: _data, xKey, yKeys, outputWindow, axisOptions, domain, domainPadding, }) => {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t;
    const data = [..._data];
    const tickValues = axisOptions === null || axisOptions === void 0 ? void 0 : axisOptions.tickValues;
    const tickDomainsX = tickValues && typeof tickValues === "object" && "x" in tickValues
        ? (0, tickHelpers_1.getDomainFromTicks)(tickValues.x)
        : (0, tickHelpers_1.getDomainFromTicks)(tickValues);
    const tickDomainsY = tickValues && typeof tickValues === "object" && "y" in tickValues
        ? (0, tickHelpers_1.getDomainFromTicks)(tickValues.y)
        : (0, tickHelpers_1.getDomainFromTicks)(tickValues);
    const isNumericalData = data.every((datum) => typeof datum[xKey] === "number");
    if (isNumericalData) {
        data.sort((a, b) => +a[xKey] - +b[xKey]);
    }
    // Input x is just extracting the xKey from each datum
    const ix = data.map((datum) => datum[xKey]);
    const ixNum = ix.map((val, i) => (isNumericalData ? val : i));
    // If user provides a domain, use that as our min / max
    // Else if, tickValues are provided, we use that instead
    // Else, we find min / max of y values across all yKeys, and use that for y range instead.
    const yMin = (_c = (_b = (_a = domain === null || domain === void 0 ? void 0 : domain.y) === null || _a === void 0 ? void 0 : _a[0]) !== null && _b !== void 0 ? _b : tickDomainsY === null || tickDomainsY === void 0 ? void 0 : tickDomainsY[0]) !== null && _c !== void 0 ? _c : Math.min(...yKeys.map((key) => {
        return data.reduce((min, curr) => {
            if (typeof curr[key] !== "number")
                return min;
            return Math.min(min, curr[key]);
        }, Infinity);
    }));
    const yMax = (_f = (_e = (_d = domain === null || domain === void 0 ? void 0 : domain.y) === null || _d === void 0 ? void 0 : _d[1]) !== null && _e !== void 0 ? _e : tickDomainsY === null || tickDomainsY === void 0 ? void 0 : tickDomainsY[1]) !== null && _f !== void 0 ? _f : Math.max(...yKeys.map((key) => {
        return data.reduce((max, curr) => {
            if (typeof curr[key] !== "number")
                return max;
            return Math.max(max, curr[key]);
        }, -Infinity);
    }));
    // Set up our y-output data structure
    const y = yKeys.reduce((acc, k) => {
        acc[k] = { i: [], o: [] };
        return acc;
    }, {});
    // Set up our y-scale, notice how domain is "flipped" because
    //  we're moving from cartesian to canvas coordinates
    // Also, if single data point, manually add upper & lower bounds so chart renders properly
    const yScaleDomain = (yMax === yMin ? [yMax + 1, yMin - 1] : [yMax, yMin]);
    const fontHeight = (_j = (_h = (_g = axisOptions === null || axisOptions === void 0 ? void 0 : axisOptions.font) === null || _g === void 0 ? void 0 : _g.getSize) === null || _h === void 0 ? void 0 : _h.call(_g)) !== null && _j !== void 0 ? _j : 0;
    // Our yScaleRange is impacted by our grid options
    const yScaleRange = (() => {
        var _a, _b, _c, _d, _e, _f;
        const xTickCount = (_b = (typeof (axisOptions === null || axisOptions === void 0 ? void 0 : axisOptions.tickCount) === "number"
            ? axisOptions === null || axisOptions === void 0 ? void 0 : axisOptions.tickCount
            : (_a = axisOptions === null || axisOptions === void 0 ? void 0 : axisOptions.tickCount) === null || _a === void 0 ? void 0 : _a.x)) !== null && _b !== void 0 ? _b : 0;
        const yLabelPosition = typeof (axisOptions === null || axisOptions === void 0 ? void 0 : axisOptions.labelPosition) === "string"
            ? axisOptions.labelPosition
            : (_c = axisOptions === null || axisOptions === void 0 ? void 0 : axisOptions.labelPosition) === null || _c === void 0 ? void 0 : _c.x;
        const xAxisSide = (_d = axisOptions === null || axisOptions === void 0 ? void 0 : axisOptions.axisSide) === null || _d === void 0 ? void 0 : _d.x;
        const yLabelOffset = (_f = (typeof (axisOptions === null || axisOptions === void 0 ? void 0 : axisOptions.labelOffset) === "number"
            ? axisOptions.labelOffset
            : (_e = axisOptions === null || axisOptions === void 0 ? void 0 : axisOptions.labelOffset) === null || _e === void 0 ? void 0 : _e.y)) !== null && _f !== void 0 ? _f : 0;
        // bottom, outset
        if (xAxisSide === "bottom" && yLabelPosition === "outset") {
            return [
                outputWindow.yMin,
                outputWindow.yMax +
                    (xTickCount > 0 ? -fontHeight - yLabelOffset * 2 : 0),
            ];
        }
        // Top outset
        if (xAxisSide === "top" && yLabelPosition === "outset") {
            return [
                outputWindow.yMin +
                    (xTickCount > 0 ? fontHeight + yLabelOffset * 2 : 0),
                outputWindow.yMax,
            ];
        }
        // Inset labels don't need added offsets
        return [outputWindow.yMin, outputWindow.yMax];
    })();
    const yScale = (0, makeScale_1.makeScale)({
        inputBounds: yScaleDomain,
        outputBounds: yScaleRange,
        isNice: true,
        padEnd: typeof domainPadding === "number" ? domainPadding : domainPadding === null || domainPadding === void 0 ? void 0 : domainPadding.bottom,
        padStart: typeof domainPadding === "number" ? domainPadding : domainPadding === null || domainPadding === void 0 ? void 0 : domainPadding.top,
    });
    yKeys.forEach((yKey) => {
        y[yKey].i = data.map((datum) => datum[yKey]);
        y[yKey].o = data.map((datum) => (typeof datum[yKey] === "number"
            ? yScale(datum[yKey])
            : datum[yKey]));
    });
    // Measure our top-most y-label if we have grid options so we can
    //  compensate for it in our x-scale.
    const topYLabel = ((_k = axisOptions === null || axisOptions === void 0 ? void 0 : axisOptions.formatYLabel) === null || _k === void 0 ? void 0 : _k.call(axisOptions, yScale.domain().at(0))) ||
        String(yScale.domain().at(0));
    // Generate our x-scale
    // If user provides a domain, use that as our min / max
    // Else if, tickValues are provided, we use that instead
    // Else, we find min / max of y values across all yKeys, and use that for y range instead.
    const ixMin = (0, asNumber_1.asNumber)((_o = (_m = (_l = domain === null || domain === void 0 ? void 0 : domain.x) === null || _l === void 0 ? void 0 : _l[0]) !== null && _m !== void 0 ? _m : tickDomainsX === null || tickDomainsX === void 0 ? void 0 : tickDomainsX[0]) !== null && _o !== void 0 ? _o : ixNum.at(0)), ixMax = (0, asNumber_1.asNumber)((_r = (_q = (_p = domain === null || domain === void 0 ? void 0 : domain.x) === null || _p === void 0 ? void 0 : _p[1]) !== null && _q !== void 0 ? _q : tickDomainsX === null || tickDomainsX === void 0 ? void 0 : tickDomainsX[1]) !== null && _r !== void 0 ? _r : ixNum.at(-1));
    const topYLabelWidth = (_t = (_s = axisOptions === null || axisOptions === void 0 ? void 0 : axisOptions.font) === null || _s === void 0 ? void 0 : _s.getTextWidth(topYLabel)) !== null && _t !== void 0 ? _t : 0;
    // Determine our x-output range based on yAxis/label options
    const oRange = (() => {
        var _a, _b, _c, _d, _e, _f;
        const yTickCount = (_b = (typeof (axisOptions === null || axisOptions === void 0 ? void 0 : axisOptions.tickCount) === "number"
            ? axisOptions === null || axisOptions === void 0 ? void 0 : axisOptions.tickCount
            : (_a = axisOptions === null || axisOptions === void 0 ? void 0 : axisOptions.tickCount) === null || _a === void 0 ? void 0 : _a.y)) !== null && _b !== void 0 ? _b : 0;
        const yLabelPosition = typeof (axisOptions === null || axisOptions === void 0 ? void 0 : axisOptions.labelPosition) === "string"
            ? axisOptions.labelPosition
            : (_c = axisOptions === null || axisOptions === void 0 ? void 0 : axisOptions.labelPosition) === null || _c === void 0 ? void 0 : _c.y;
        const yAxisSide = (_d = axisOptions === null || axisOptions === void 0 ? void 0 : axisOptions.axisSide) === null || _d === void 0 ? void 0 : _d.y;
        const yLabelOffset = (_f = (typeof (axisOptions === null || axisOptions === void 0 ? void 0 : axisOptions.labelOffset) === "number"
            ? axisOptions.labelOffset
            : (_e = axisOptions === null || axisOptions === void 0 ? void 0 : axisOptions.labelOffset) === null || _e === void 0 ? void 0 : _e.y)) !== null && _f !== void 0 ? _f : 0;
        // Left axes, outset label
        if (yAxisSide === "left" && yLabelPosition === "outset") {
            return [
                outputWindow.xMin +
                    (yTickCount > 0 ? topYLabelWidth + yLabelOffset : 0),
                outputWindow.xMax,
            ];
        }
        // Right axes, outset label
        if (yAxisSide === "right" && yLabelPosition === "outset") {
            return [
                outputWindow.xMin,
                outputWindow.xMax +
                    (yTickCount > 0 ? -topYLabelWidth - yLabelOffset : 0),
            ];
        }
        // Inset labels don't need added offsets
        return [outputWindow.xMin, outputWindow.xMax];
    })();
    const xScale = (0, makeScale_1.makeScale)({
        // if single data point, manually add upper & lower bounds so chart renders properly
        inputBounds: ixMin === ixMax ? [ixMin - 1, ixMax + 1] : [ixMin, ixMax],
        outputBounds: oRange,
        padStart: typeof domainPadding === "number" ? domainPadding : domainPadding === null || domainPadding === void 0 ? void 0 : domainPadding.left,
        padEnd: typeof domainPadding === "number" ? domainPadding : domainPadding === null || domainPadding === void 0 ? void 0 : domainPadding.right,
    });
    const ox = ixNum.map((x) => xScale(x));
    return {
        ix,
        ox,
        y,
        xScale,
        yScale,
        isNumericalData,
    };
};
exports.transformInputData = transformInputData;
