// zeego/lib/module/context-menu/context-menu.js
import React from 'react';
import { create, MenuDisplayName } from '../menu';

const Root = _ref => {
  let {
    children
  } = _ref;
  return /*#__PURE__*/React.createElement(React.Fragment, null, children);
};

Root.displayName = MenuDisplayName.Root;

const Trigger = _ref2 => {
  let {
    children
  } = _ref2;
  return /*#__PURE__*/React.createElement(React.Fragment, null, children);
};

Trigger.displayName = MenuDisplayName.Trigger;

const Content = _ref3 => {
  let {
    children
  } = _ref3;
  return /*#__PURE__*/React.createElement(React.Fragment, null, children);
};

Content.displayName = MenuDisplayName.Content;

const Item = _ref4 => {
  let {
    children
  } = _ref4;
  return /*#__PURE__*/React.createElement(React.Fragment, null, children);
};

Item.displayName = MenuDisplayName.Item;

const ItemIcon = () => {
  return /*#__PURE__*/React.createElement(React.Fragment, null);
};

ItemIcon.displayName = MenuDisplayName.ItemIcon;

const ItemImage = () => {
  return /*#__PURE__*/React.createElement(React.Fragment, null);
};

ItemImage.displayName = MenuDisplayName.ItemImage;

const SubTrigger = _ref5 => {
  let {
    children
  } = _ref5;
  return /*#__PURE__*/React.createElement(React.Fragment, null, children);
};

SubTrigger.displayName = MenuDisplayName.SubTrigger;

const ItemTitle = _ref6 => {
  let {
    children
  } = _ref6;
  return /*#__PURE__*/React.createElement(React.Fragment, null, children);
};

ItemTitle.displayName = MenuDisplayName.ItemTitle;

const ItemSubtitle = _ref7 => {
  let {
    children
  } = _ref7;
  return /*#__PURE__*/React.createElement(React.Fragment, null, children);
};

ItemSubtitle.displayName = MenuDisplayName.ItemSubtitle;

const Group = _ref8 => {
  let {
    children
  } = _ref8;
  return /*#__PURE__*/React.createElement(React.Fragment, null, children);
};

Group.displayName = MenuDisplayName.Group;

const Separator = () => {
  return /*#__PURE__*/React.createElement(React.Fragment, null);
};

Separator.displayName = MenuDisplayName.Separator;

const CheckboxItem = () => /*#__PURE__*/React.createElement(React.Fragment, null);

CheckboxItem.displayName = MenuDisplayName.CheckboxItem;

const ItemIndicator = () => /*#__PURE__*/React.createElement(React.Fragment, null);

ItemIndicator.displayName = MenuDisplayName.ItemIndicator;

const Label = () => /*#__PURE__*/React.createElement(React.Fragment, null);

Label.displayName = MenuDisplayName.Label;

const Preview = () => /*#__PURE__*/React.createElement(React.Fragment, null);

Preview.displayName = MenuDisplayName.Preview;
const Arrow = create(_ => /*#__PURE__*/React.createElement(React.Fragment, null), 'Arrow');
const Sub = create(_ => /*#__PURE__*/React.createElement(React.Fragment, null), 'Sub');
const SubContent = create(_ => /*#__PURE__*/React.createElement(React.Fragment, null), 'SubContent');
const Auxiliary = create(_ => /*#__PURE__*/React.createElement(React.Fragment, null), 'Auxiliary');
export { Root, Trigger, Content, Item, ItemTitle, ItemSubtitle, SubTrigger, Group, ItemIcon, Separator, CheckboxItem, ItemIndicator, ItemImage, Label, Preview, Arrow, Sub, SubContent, Auxiliary };
