// zeego/lib/commonjs/context-menu/context-menu.ios.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Trigger = exports.SubTrigger = exports.SubContent = exports.Sub = exports.Separator = exports.Root = exports.Preview = exports.Label = exports.ItemTitle = exports.ItemSubtitle = exports.ItemIndicator = exports.ItemImage = exports.ItemIcon = exports.Item = exports.Group = exports.Content = exports.CheckboxItem = exports.Auxiliary = exports.Arrow = void 0;

var _menu = require("../menu");

const {
  Root,
  Trigger,
  Content,
  Item,
  ItemTitle,
  ItemSubtitle,
  SubTrigger,
  Group,
  CheckboxItem,
  ItemIcon,
  Separator,
  ItemIndicator,
  ItemImage,
  Label,
  Preview,
  Arrow,
  Sub,
  SubContent,
  Auxiliary
} = (0, _menu.createIosMenu)('ContextMenu');
exports.Auxiliary = Auxiliary;
exports.SubContent = SubContent;
exports.Sub = Sub;
exports.Arrow = Arrow;
exports.Preview = Preview;
exports.Label = Label;
exports.ItemImage = ItemImage;
exports.ItemIndicator = ItemIndicator;
exports.Separator = Separator;
exports.ItemIcon = ItemIcon;
exports.CheckboxItem = CheckboxItem;
exports.Group = Group;
exports.SubTrigger = SubTrigger;
exports.ItemSubtitle = ItemSubtitle;
exports.ItemTitle = ItemTitle;
exports.Item = Item;
exports.Content = Content;
exports.Trigger = Trigger;
exports.Root = Root;
