// zeego/lib/commonjs/context-menu/context-menu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Trigger = exports.SubTrigger = exports.SubContent = exports.Sub = exports.Separator = exports.Root = exports.Preview = exports.Label = exports.ItemTitle = exports.ItemSubtitle = exports.ItemIndicator = exports.ItemImage = exports.ItemIcon = exports.Item = exports.Group = exports.Content = exports.CheckboxItem = exports.Auxiliary = exports.Arrow = void 0;

var _react = _interopRequireDefault(require("react"));

var _menu = require("../menu");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const Root = _ref => {
  let {
    children
  } = _ref;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, children);
};

exports.Root = Root;
Root.displayName = _menu.MenuDisplayName.Root;

const Trigger = _ref2 => {
  let {
    children
  } = _ref2;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, children);
};

exports.Trigger = Trigger;
Trigger.displayName = _menu.MenuDisplayName.Trigger;

const Content = _ref3 => {
  let {
    children
  } = _ref3;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, children);
};

exports.Content = Content;
Content.displayName = _menu.MenuDisplayName.Content;

const Item = _ref4 => {
  let {
    children
  } = _ref4;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, children);
};

exports.Item = Item;
Item.displayName = _menu.MenuDisplayName.Item;

const ItemIcon = () => {
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null);
};

exports.ItemIcon = ItemIcon;
ItemIcon.displayName = _menu.MenuDisplayName.ItemIcon;

const ItemImage = () => {
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null);
};

exports.ItemImage = ItemImage;
ItemImage.displayName = _menu.MenuDisplayName.ItemImage;

const SubTrigger = _ref5 => {
  let {
    children
  } = _ref5;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, children);
};

exports.SubTrigger = SubTrigger;
SubTrigger.displayName = _menu.MenuDisplayName.SubTrigger;

const ItemTitle = _ref6 => {
  let {
    children
  } = _ref6;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, children);
};

exports.ItemTitle = ItemTitle;
ItemTitle.displayName = _menu.MenuDisplayName.ItemTitle;

const ItemSubtitle = _ref7 => {
  let {
    children
  } = _ref7;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, children);
};

exports.ItemSubtitle = ItemSubtitle;
ItemSubtitle.displayName = _menu.MenuDisplayName.ItemSubtitle;

const Group = _ref8 => {
  let {
    children
  } = _ref8;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, children);
};

exports.Group = Group;
Group.displayName = _menu.MenuDisplayName.Group;

const Separator = () => {
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null);
};

exports.Separator = Separator;
Separator.displayName = _menu.MenuDisplayName.Separator;

const CheckboxItem = () => /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null);

exports.CheckboxItem = CheckboxItem;
CheckboxItem.displayName = _menu.MenuDisplayName.CheckboxItem;

const ItemIndicator = () => /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null);

exports.ItemIndicator = ItemIndicator;
ItemIndicator.displayName = _menu.MenuDisplayName.ItemIndicator;

const Label = () => /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null);

exports.Label = Label;
Label.displayName = _menu.MenuDisplayName.Label;

const Preview = () => /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null);

exports.Preview = Preview;
Preview.displayName = _menu.MenuDisplayName.Preview;
const Arrow = (0, _menu.create)(_ => /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null), 'Arrow');
exports.Arrow = Arrow;
const Sub = (0, _menu.create)(_ => /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null), 'Sub');
exports.Sub = Sub;
const SubContent = (0, _menu.create)(_ => /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null), 'SubContent');
exports.SubContent = SubContent;
const Auxiliary = (0, _menu.create)(_ => /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null), 'Auxiliary');
exports.Auxiliary = Auxiliary;
