// @shopify/react-native-skia/lib/commonjs/external/reanimated/moduleWrapper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.stopMapper = exports.startMapper = exports.runOnUI = exports.runOnJS = exports.isSharedValue = exports.HAS_REANIMATED3 = exports.HAS_REANIMATED2 = void 0;
exports.throwOnMissingReanimated = throwOnMissingReanimated;
exports.useSharedValue = exports.useFrameCallback = exports.useDerivedValue = exports.useAnimatedReaction = void 0;
var _Reanimated, _Reanimated2, _Reanimated3, _Reanimated4, _Reanimated5, _Reanimated6, _Reanimated7, _Reanimated8;
/* eslint-disable @typescript-eslint/no-explicit-any */

// This one is needed for the deprecated useSharedValue function
// We can remove it once we remove the deprecation

let Reanimated2;
let Reanimated3;
let reanimatedVersion;
try {
  Reanimated2 = require("react-native-reanimated");
  reanimatedVersion = require("react-native-reanimated/package.json").version;
  if (reanimatedVersion && (reanimatedVersion >= "3.0.0" || reanimatedVersion.includes("3.0.0-"))) {
    Reanimated3 = Reanimated2;
  }
} catch (e) {}
const HAS_REANIMATED2 = exports.HAS_REANIMATED2 = !!Reanimated2;
const HAS_REANIMATED3 = exports.HAS_REANIMATED3 = !!Reanimated3;
function throwOnMissingReanimated() {
  if (!HAS_REANIMATED2) {
    throw new Error("Reanimated was not found, make sure react-native-reanimated package is installed if you want to use \
      react-native-skia's integration layer API.");
  }
}
const useSharedValue = exports.useSharedValue = ((_Reanimated = Reanimated2) === null || _Reanimated === void 0 ? void 0 : _Reanimated.useSharedValue) || throwOnMissingReanimated;
const useDerivedValue = exports.useDerivedValue = ((_Reanimated2 = Reanimated2) === null || _Reanimated2 === void 0 ? void 0 : _Reanimated2.useDerivedValue) || throwOnMissingReanimated;
const useFrameCallback = exports.useFrameCallback = ((_Reanimated3 = Reanimated2) === null || _Reanimated3 === void 0 ? void 0 : _Reanimated3.useFrameCallback) || throwOnMissingReanimated;
const startMapper = exports.startMapper = ((_Reanimated4 = Reanimated2) === null || _Reanimated4 === void 0 ? void 0 : _Reanimated4.startMapper) || throwOnMissingReanimated;
const stopMapper = exports.stopMapper = ((_Reanimated5 = Reanimated2) === null || _Reanimated5 === void 0 ? void 0 : _Reanimated5.stopMapper) || throwOnMissingReanimated;
const runOnJS = exports.runOnJS = ((_Reanimated6 = Reanimated2) === null || _Reanimated6 === void 0 ? void 0 : _Reanimated6.runOnJS) || throwOnMissingReanimated;
const runOnUI = exports.runOnUI = ((_Reanimated7 = Reanimated2) === null || _Reanimated7 === void 0 ? void 0 : _Reanimated7.runOnUI) || throwOnMissingReanimated;
const useAnimatedReaction = exports.useAnimatedReaction = ((_Reanimated8 = Reanimated2) === null || _Reanimated8 === void 0 ? void 0 : _Reanimated8.useAnimatedReaction) || throwOnMissingReanimated;
const isSharedValue = value => {
  return !!value && (Reanimated3 ? Reanimated3.isSharedValue(value) : value.value !== undefined);
};
exports.isSharedValue = isSharedValue;
