// zeego/lib/commonjs/menu/display-names.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.menuify = exports.create = exports.MenuDisplayName = void 0;

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const MenuDisplayName = {
  Root: 'Root',
  Item: 'Item',
  Content: 'Content',
  Separator: 'Separator',
  Group: 'Group',
  ItemTitle: 'ItemTitle',
  ItemSubtitle: 'ItemSubtitle',
  ItemIcon: 'ItemIcon',
  ItemImage: 'ItemImage',
  Trigger: 'Trigger',
  SubTrigger: 'SubTrigger',
  Sub: 'Sub',
  SubContent: 'SubContent',
  CheckboxItem: 'CheckboxItem',
  Label: 'Label',
  ItemIndicator: 'ItemIndicator',
  Preview: 'Preview',
  Arrow: 'Arrow',
  Auxiliary: 'Auxiliary'
};
exports.MenuDisplayName = MenuDisplayName;

/**
 * @deprecated Please use `create` instead.
 */
const menuify = (Component, displayName) => {
  const MenuComponent = props => {
    return /*#__PURE__*/_react.default.createElement(Component, props);
  };

  MenuComponent.displayName = displayName;
  return MenuComponent;
};

exports.menuify = menuify;

const create = (Component, displayName) => {
  const MenuComponent = props => {
    return /*#__PURE__*/_react.default.createElement(Component, props);
  };

  MenuComponent.displayName = displayName;
  return MenuComponent;
};

exports.create = create;
