// react-native-ios-context-menu/build/constants/Enums.js
/** Used for `IconConfigKeys.iconType` or `IconConfigKeys.iconType` */
export const IconTypes = {
    'NONE': 'NONE',
    'SYSTEM': 'SYSTEM',
    'ASSET': 'ASSET',
    'REQUIRE': 'REQUIRE',
};
/** Used for `MenuConfigKeys.menuOptions` */
export const MenuOptions = {
    destructive: 'destructive',
    displayInline: 'displayInline',
};
/** Used for `MenuActionKeys.menuAttributes` */
export const MenuElementAtrributes = {
    hidden: 'hidden',
    disabled: 'disabled',
    destructive: 'destructive',
};
/** Used for `MenuActionKeys.menuState` */
export const MenuElementState = {
    on: 'on',
    off: 'off',
    mixed: 'mixed',
};
/** Used for `PreviewConfigKeys.previewType` */
export const PreviewType = {
    'DEFAULT': 'DEFAULT',
    'CUSTOM': 'CUSTOM',
};
/** Used for `PreviewConfigKeys.previewSize` */
export const PreviewSize = {
    'INHERIT': 'INHERIT',
    'STRETCH': 'STRETCH',
};
/** Used for `PreviewConfigKeys.preferredCommitStyle` */
export const CommitStyle = {
    'pop': 'pop',
    'dismiss': 'dismiss',
};
export const IconConfigKeys = {
    iconType: 'iconType',
    iconValue: 'iconValue',
    iconTint: 'iconTint',
};
export const PreviewConfigKeys = {
    previewType: 'previewType',
    previewSize: 'previewSize',
    isResizeAnimated: 'isResizeAnimated',
    borderRadius: 'borderRadius',
    backgroundColor: 'backgroundColor',
    preferredCommitStyle: 'preferredCommitStyle', // optional - CommitStyle value
};
export const MenuActionKeys = {
    actionKey: 'actionKey',
    actionTitle: 'actionTitle',
    icon: 'icon',
    menuState: 'menuState',
    menuAttributes: 'menuAttributes',
    discoverabilityTitle: 'discoverabilityTitle', // optional - string value
};
export const MenuConfigKeys = {
    menuTitle: 'menuTitle',
    menuOptions: 'menuOptions',
    icon: 'icon',
    menuItems: 'menuItems', // optional - array of MenuConfigKeys/MenuActionKeys
};
